port awx to run natively on python3.6+

ryanpetrello opened the pull request:
This is still a work in progress - our plan is to continue to test and rebase this in the coming weeks.

It is *very* likely that this is still not fully compatible in its current state, but it _does_ pass known unit/functional tests that we have.

More testing to come...

ryanpetrello committed:
switch image builds to py3

ryanpetrello committed:
convert py2 -> py3

ryanpetrello commented:
recheck

ryanpetrello commented:
recheck

Spredzy commented:
recheck

ryanpetrello commented:
recheck

ryanpetrello committed:
parametrize PYTHON for Ubuntu py35 support

ryanpetrello committed:
remove dm.xmlsec.binding

python-saml uses dm.xmlsec.binding only supports python2
by moving to py3, we now use python3-saml (which uses python-xmlsec
instead)

see: https://github.com/onelogin/python-saml/issues/145#issuecomment-222021691

Commit:
fix some isolated py3 bugs

ryanpetrello committed:
make py3 packaging work for k8s

AlanCoding commented on awx/main/management/commands/inventory_import.py:
```diff
@@ -155,6 +155,8 @@ def command_to_json(self, cmd):
 
         proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env)
         stdout, stderr = proc.communicate()
```
Earlier above in this file we have

```python
        env.pop('PYTHONPATH', None)  # default to none if no python_ver matches
        if os.path.isdir(os.path.join(venv_libdir, "python2.7")):
            env['PYTHONPATH'] = os.path.join(venv_libdir, "python2.7", "site-packages") + ":"
```

I fear this is going to have some majorly negative effects on the users who have a python3 venv installed, although I do not yet fully understand the mechanics of that.

It's something that I'm digging into a lot, because of the work on inventory plugins. We are going in the direction of allowing use of organization custom venv for inventory imports. That should open up some means of passing through the details. Until then, I figure this will be a major noted known issue that needs to be resolved after merge.

ryanpetrello replied:
Yep, this is a thing in my brain that needs solving, but I don't intend to solve it in this PR.

We'll need some changes here (and a few other places) when we support ansible venvs in py3 (which this PR doesn't totally do quite yet; it only makes awx _itself_ use py3).

elyezer commented on Makefile:
```diff
@@ -195,7 +191,7 @@ version_file:
 		. $(VENV_BASE)/awx/bin/activate; \
 	fi; \
-	python -c "import awx as awx; print awx.__version__" > /var/lib/awx/.awx_version; \
+	python -c "import awx as awx; print(awx.__version__)" > /var/lib/awx/.awx_version; \
```
Isn't `import awx as awx` redundant? Maybe it could be just `import awx`, doesn't it?

ryanpetrello replied:
Ha, good eye :)

If I had to guess, this is probably a leftover from our `tower -> awx` renaming when we open sourced.

elyezer commented on awx/api/filters.py:
```diff
@@ -48,7 +48,7 @@ class TypeFilterBackend(BaseFilterBackend):
         try:
             types = None
-            for key, value in request.query_params.items():
+            for key, value in list(request.query_params.items()):
```
I don't think we need to cast it to a list:

```
In [5]: d['a'] = 1                                                                                                 

In [6]: for k, v in d.items(): 
   ...:     print(k, v) 
   ...:                                                                                                            
a 1
```

ryanpetrello replied:
Agreed, looks like I got over-aggressive with some `sed` here.

elyezer commented on awx/api/filters.py:
```diff
@@ -65,7 +65,7 @@ def filter_queryset(self, request, queryset, view):
                 model_type = get_type_for_model(model)
                 if 'polymorphic_ctype' in get_all_field_names(model):
-                    types_pks = set([v for k,v in types_map.items() if k in types])
+                    types_pks = set([v for k,v in list(types_map.items()) if k in types])
```
This could be shortened to `types_pks = set(types_map.values())`, for example:

```python
In [8]: set(d.values())                                                                                            
Out[8]: {1}

In [9]: set([v for k, v in d.items()])                                                                             
Out[9]: {1}
```

elyezer commented on awx/api/filters.py:
```diff
@@ -363,12 +363,12 @@ def filter_queryset(self, request, queryset, view):
                 if search_filters and search_filter_relation == 'OR':
                     q = Q()
-                    for term, constrains in search_filters.iteritems():
+                    for term, constrains in list(search_filters.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/filters.py:
```diff
@@ -363,12 +363,12 @@ def filter_queryset(self, request, queryset, view):
                     args.append(q)
                 elif search_filters and search_filter_relation == 'AND':
-                    for term, constrains in search_filters.iteritems():
+                    for term, constrains in list(search_filters.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/filters.py:
```diff
@@ -395,7 +395,7 @@ class OrderByBackend(BaseFilterBackend):
         try:
             order_by = None
-            for key, value in request.query_params.items():
+            for key, value in list(request.query_params.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/generics.py:
```diff
@@ -347,7 +348,7 @@ def get_serializer(self, *args, **kwargs):
         if hasattr(self, '_raw_data_form_marker'):
             # Always remove read only fields from serializer.
-            for name, field in serializer.fields.items():
+            for name, field in list(serializer.fields.items()):
```
I think we can drop the cast to `list`

ryanpetrello replied:
`serializer.fields` is modified in place lines later.

elyezer commented on awx/api/generics.py:
```diff
@@ -870,7 +871,7 @@ def _build_create_dict(self, obj):
                 self._get_copy_return_serializer(), method='POST'
             )
-            for field_name, field_info in all_fields.items():
+            for field_name, field_info in list(all_fields.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/metadata.py:
```diff
@@ -157,7 +157,7 @@ def determine_actions(self, request, view):
                 view.request = request
 
-            for field, meta in actions[method].items():
+            for field, meta in list(actions[method].items()):
```
I think we can drop the cast to `list`

ryanpetrello replied:
Same, `actions` is modified in place later in this block.

elyezer commented on awx/api/serializers.py:
```diff
@@ -203,11 +203,11 @@ class Meta:
     @staticmethod
     def _is_extra_kwargs(x):
-        return isinstance(x, dict) and all([isinstance(k, basestring) and isinstance(v, dict) for k,v in x.items()])
+        return isinstance(x, dict) and all([isinstance(k, str) and isinstance(v, dict) for k,v in list(x.items())])
```
I think we can drop the cast to `list`

elyezer commented on awx/api/serializers.py:
```diff
@@ -240,9 +240,9 @@ def _update_meta(cls, base, meta, other=None):
                 new_val = {}
                 if base:
-                    for k,v in meta_val.items():
+                    for k,v in list(meta_val.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/serializers.py:
```diff
@@ -240,9 +240,9 @@ def _update_meta(cls, base, meta, other=None):
+                    for k,v in list(meta_val.items()):
                         new_val[k] = copy.deepcopy(v)
-                for k,v in val.items():
+                for k,v in list(val.items()):
```
I think we can drop the cast to `list`

ryanpetrello replied:
😂

elyezer commented on awx/api/serializers.py:
```diff
@@ -366,7 +364,7 @@ def get_summary_fields(self, obj):
         # displaying lists of items without additional API requests.
         summary_fields = OrderedDict()
-        for fk, related_fields in SUMMARIZABLE_FK_FIELDS.items():
+        for fk, related_fields in list(SUMMARIZABLE_FK_FIELDS.items()):
```
I think we can drop the cast to `list`

elyezer commented on awx/api/serializers.py:
```diff
@@ -565,7 +563,7 @@ def get_validation_exclusions(self, obj=None):
         opts = cls._meta.concrete_model._meta
         exclusions = [field.name for field in opts.fields]
-        for field_name, field in self.fields.items():
+        for field_name, field in list(self.fields.items()):
```
I think we can drop the cast to `list`

PS.: Review the rest of the changes for this :smile:

elyezer commented on awx/conf/views.py:
```diff
@@ -72,7 +72,7 @@ class SettingSingletonDetail(RetrieveUpdateDestroyAPIView):
     def get_queryset(self):
         self.category_slug = self.kwargs.get('category_slug', 'all')
-        all_category_slugs = settings_registry.get_registered_categories(features_enabled=get_licensed_features()).keys()
+        all_category_slugs = list(settings_registry.get_registered_categories(features_enabled=get_licensed_features()).keys())
```
I think we can drop the cast to `list`

PS: adding this here for a different pattern other them casting the return value of `items`

ryanpetrello replied:
Later in this function, we modify this data structure as a list:
```
>>> x = {}.keys()
>>> x.remove
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'dict_keys' object has no attribute 'remove'
```

elyezer commented on awx/lib/awx_display_callback/events.py:
```diff
@@ -165,11 +162,11 @@ def get_end_dict(self):
 
     def dump(self, fileobj, data, max_width=78, flush=False):
-        b64data = base64.b64encode(json.dumps(data))
+        b64data = base64.b64encode(json.dumps(data).encode('utf-8')).decode()
```
I don't know much about the code, but just checking the return of `dumps`, it seems the encode is not necessary since it will probably return an UTF-8 string:

```python
In [26]: r = json.dumps({"name": "elyézer"})                                                                       

In [27]: r                                                                                                         
Out[27]: '{"name": "ely\\u00e9zer"}'

In [28]: r.encode()  # utf-8 is the default value here                                                                                                
Out[28]: b'{"name": "ely\\u00e9zer"}'
```

ryanpetrello replied:
Maybe?  For what it's worth, this portion of code was lifted from ansible-runner, which we'll be switching to in the near future anyways:

https://github.com/ansible/ansible-runner/blob/master/ansible_runner/display_callback/events.py#L163

elyezer replied:
I see, in that case it may be necessary for some cases. Thank you for providing more information.

elyezer commented on awx/main/access.py:
```diff
@@ -1397,6 +1397,8 @@ def changes_are_non_sensitive(self, obj, data):
         ]
 
         for k, v in data.items():
+            if k not in [x.name for x in obj._meta.concrete_fields]:
```
Depending on the length of `data.items()` and `obj._meta_concrete_fields` this can be a very time expensive task since the list will have to be evaluated over and over again. If you think this is going to be an issue, then you could create a cache. Something like:

```python
obj_concrete_fields = [x.name for x in obj._meta.concrete_fields]
for k, v in data.items():
    if k not in obj_concrete_fields:
        continue
    # ...
```

ryanpetrello replied:
I don't expect any notable performance issues here; we're generally talking about the number of fields on a given model vs keys in the data payload.  _Maybe_ `N*M = ~1000`?

This isn't a crazy common request - it's used (in part) to determine RBAC-wise if a user can make changes to a Job Template.

elyezer replied:
I see, thanks for providing more context. About 1000 won't hurt.

AlanCoding replied:
shell_plus

```python
print [len(cls._meta.concrete_fields) for cls in locals().values() if inspect.isclass(cls) and issubclass(cls, Model) and hasattr(cls, '_meta')]
[11, 51, 6, 4, 13, 17, 18, 4, 10, 6, 11, 5, 18, 28, 19, 11, 3, 3, 8, 5, 9, 8, 6, 26, 11, 4, 2, 54, 4, 45, 41, 18, 6, 12, 3, 2, 11, 13, 15, 11, 16, 7, 10, 5, 33, 5, 8, 2, 40, 12, 5, 6, 22, 35, 3, 9, 29, 11, 38, 14, 16, 36, 3, 3, 3]
```

max 54 -> so about 3,000

AlanCoding replied:
but it would find its match half-way through on average, so 1,500

elyezer commented on awx/main/expect/isolated_manager.py:
```diff
@@ -246,7 +250,7 @@ def build_isolated_job_data(self):
 
     def _missing_artifacts(self, path_list, output):
-        missing_artifacts = filter(lambda path: not os.path.exists(path), path_list)
+        missing_artifacts = list(filter(lambda path: not os.path.exists(path), path_list))
```
`filter` already return an iterator. Maybe we can drop the list cast here.

ryanpetrello replied:
There are len/truthiness checks of this var later in code, so it was easier to just make it a list.

elyezer commented on awx/main/fields.py:
```diff
@@ -514,10 +514,10 @@ def schema(self, model_instance):
             if field['type'] == 'become_method':
                 field.pop('type')
-                field['choices'] = map(operator.itemgetter(0), CHOICES_PRIVILEGE_ESCALATION_METHODS)
+                field['choices'] = list(map(operator.itemgetter(0), CHOICES_PRIVILEGE_ESCALATION_METHODS))
```
`map` returns an iterator. Maybe the cast to list can be dropped

PS: worth checking for more like this

ryanpetrello replied:
This was a case where our DRF code was doing list indexing and expected a list.

elyezer commented on awx/main/models/ad_hoc_commands.py:
```diff
@@ -109,15 +109,15 @@ def clean_limit(self):
 
     def clean_module_name(self):
-        if type(self.module_name) not in (str, unicode):
+        if type(self.module_name) is not str:
```
Do you think that `if not isinstance(self.module_name, str)` is cleaner?

elyezer commented on awx/main/models/ad_hoc_commands.py:
```diff
@@ -109,15 +109,15 @@ def clean_limit(self):
 
     def clean_module_args(self):
-        if type(self.module_args) not in (str, unicode):
+        if type(self.module_args) is not str:
```
Same here, see https://github.com/ansible/awx/pull/2538/files#r246072529

elyezer commented on awx/main/tasks.py:
```diff
@@ -722,12 +722,12 @@ def build_private_data_files(self, instance, **kwargs):
             ssh_too_old = True if ssh_ver == "unknown" else Version(ssh_ver) < Version("6.0")
             openssh_keys_supported = ssh_ver != "unknown" and Version(ssh_ver) >= Version("6.5")
-            for credential, data in private_data.get('credentials', {}).iteritems():
+            for credential, data in list(private_data.get('credentials', {}).items()):
```
This is casting to list but `items` is being called. `list` object does not have the `items` method.

elyezer commented on awx/main/tasks.py:
```diff
@@ -722,12 +722,12 @@ def build_private_data_files(self, instance, **kwargs):
                 if 'OPENSSH PRIVATE KEY' in data and not openssh_keys_supported:
                     raise RuntimeError(OPENSSH_KEY_ERROR)
-            for credential, data in private_data.get('credentials', {}).iteritems():
+            for credential, data in list(private_data.get('credentials', {}).items()):
```
This is casting to list but `items` is being called. `list` object does not have the `items` method.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -3,8 +3,7 @@
 
```
For multiple types of jobs, we call `add_awx_venv`, notably, inventory updates and system jobs.

Since the Ansible venv this is supporting is python 2.7... might that fail some weird scenarios?

Maybe adding the AWX venv to these things _shouldn't_ happen.

AlanCoding commented on awx/main/expect/run.py:
```diff
@@ -97,14 +96,8 @@ def run_pexpect(args, cwd, env, logfile,
-    # https://github.com/pexpect/pexpect/issues/512
-    for k, v in env.items():
-        if isinstance(v, six.text_type):
-            env[k] = v.encode('utf-8')
```
are we sure that the values are going to be properly strings and not bytes? Does this change leave us trusting the source of these? Couldn't that still go wrong?

ryanpetrello replied:
Nope, this specific code block isn't necessary in python3 because strings are unicode by default.

The underlying bug in pexpect doesn't show up in python3:

```
~ cat hang.py && python hang.py
# -*- coding: utf-8 -*-

import pexpect
pexpect.spawn(
    'sleep', ['1'], cwd='/tmp/', env={'FOO': u'Iñtërnâtiônàlizætiøn'}, ignore_sighup=True,
    encoding='utf-8', echo=False
)
```

(it hangs)

```
~ python3 hang.py && echo $?
0
```

(it doesn't hang)

...so in py3, we just don't need this code anymore.

ryanpetrello replied:
![leadingajarantelopegroundsquirrel-size_restricted](https://user-images.githubusercontent.com/214912/52966934-8dd1b880-3376-11e9-9312-18553ed93f7e.gif)

Nope, I've totally got this wrong.  We still need this hack in py3.  I'm going to get this fixed in pexpect, where it belongs.

AlanCoding commented:
I believe I introduced a conflict in the Dockerfile by changing some things about how the development supervisor config file was set up.

ryanpetrello commented:
@AlanCoding I just rebased and pushed, we should be good now.

AlanCoding commented on requirements/requirements_ansible.in:
```diff
@@ -45,7 +45,6 @@ psutil==5.4.3    # same as AWX requirement
 setuptools==36.0.1
 pip==9.0.1
 # VMware
-psphere==0.5.2
```
I tried to cull this list before, and if something in is the current list, it's probably because something prevented me from getting rid of it. It looks like psphere is still imported in the vmware script.

https://github.com/ansible/ansible/blob/a11f631ee414ba1048b04564cc901f1ff8b67479/contrib/inventory/vmware.py

However, our script is called `vmware_inventory.py` in plugins. The documentation in these scripts are not helping me understand what happened there, but it looks like maybe this really was just missed?

ryanpetrello replied:
Maybe @wenottingham's assertions at https://github.com/ansible/awx/pull/2506 are not actually true?

ryanpetrello replied:
Also, I'm mostly certain that the python vmware psphere SDK doesn’t support py3:

https://github.com/jkinred/psphere/issues/9

My understanding of this package is the same as @wenottingham, that it's dead, and we tell people to stop using it.

If a user was _really_ wanting to use this library in py2 ansible world, they _could_ make a custom venv that it installed it; but our plan is not to bring it forward in py2/3 world.

AlanCoding replied:
yeah, this looks fine to me. It was planned to be removed with this PR. I don't anticipate any issues.

AlanCoding replied:
Thankfully, it looks like the plugin is a continuation of `vmware_inventory.py` and not `vmware.py`, and also does not use psphere.

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/vmware_vm_inventory.py

So we should be safe there as well.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -831,7 +831,7 @@ def build_inventory(self, instance, **kwargs):
         handle, path = tempfile.mkstemp(dir=kwargs.get('private_data_dir', None))
         f = os.fdopen(handle, 'w')
-        f.write('#! /usr/bin/env python\n# -*- coding: utf-8 -*-\nprint %r\n' % json_data)
+        f.write('#! /usr/bin/env python\n# -*- coding: utf-8 -*-\nprint(%r)\n' % json_data)
```
Does `print('foo')` work on all python2 that we might encounter?

ryanpetrello replied:
Yes, `print()` was introduced in python2.6.

shanemcd committed:
Enable py3 SCL if needed

shanemcd committed:
Dependency updates for Python 3

ryanpetrello committed:
support isolated runs in py2 *and* py3 (for now)

once we merge in runner support for isolated environments, we can
revert this commit (because we'll always run isolated code using python3
executables)

ryanpetrello committed:
fix a few UTF-8 bugs on Ubuntu related to stdout text downloads

shanemcd committed:
Python 3 / Upstream Kubernetes

AlanCoding commented on awx/conf/signals.py:
```diff
@@ -9,15 +9,11 @@
 logger = logging.getLogger('awx.conf.signals')
 
-awx.main.signals.model_serializer_mapping[Setting] = SettingSerializer
-
```
👍

AlanCoding commented on awx/lib/awx_display_callback/events.py:
```diff
@@ -56,9 +54,8 @@ def set(self, key, value):
-        partial_data = json.dumps(value)
         with os.fdopen(os.open(write_location, os.O_WRONLY | os.O_CREAT, stat.S_IRUSR | stat.S_IWUSR), 'w') as f:
-            f.write(partial_data)
+            f.write(value)
```
so I guess this was already dumped before we got here. That looks like it lines up with the use of the method in this file.

ryanpetrello committed:
fix a few isolated issues related to the py2 -> py3 move

ryanpetrello committed:
work around a py3 bug in celerybeat

ryanpetrello commented:
regate
